iroha_futures/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+//! Crate with various iroha futures
+
+use std::{
+    convert::TryFrom,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+
+pub use futures_derive::*;
+use iroha_logger::telemetry::{Telemetry, TelemetryFields};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+/// Future which sends info with telemetry about number and length of polls
+#[derive(Debug, Clone, Copy)]
+pub struct TelemetryFuture<F> {
+    future: F,
+    id: u64,
+    name: &'static str,
+}
+
+impl<F> TelemetryFuture<F> {
+    /// Constructor for future
+    pub fn new(future: F, name: &'static str) -> Self {
+        let id = rand::random();
+        Self { future, name, id }
+    }
+}
+
+/// Telemetry info for future polling
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct FuturePollTelemetry {
+    /// Future id
+    pub id: u64,
+    /// Future name
+    pub name: String,
+    /// Duration of poll
+    pub duration: Duration,
+}
+
+/// Error which happends during conversion from telemetry
+#[derive(Debug, Clone, Copy)]
+pub struct FromTelemetryError;
+
+impl TryFrom<&Telemetry> for FuturePollTelemetry {
+    type Error = FromTelemetryError;
+
+    #[allow(clippy::unwrap_in_result, clippy::unwrap_used)]
+    fn try_from(Telemetry { target, fields }: &Telemetry) -> Result<Self, FromTelemetryError> {
+        if target != &"iroha_futures" && fields.len() != 3 {
+            return Err(FromTelemetryError);
+        }
+
+        let TelemetryFields(fields) = fields;
+        let (mut id, mut name, mut duration) = (None, None, None);
+
+        for field in fields {
+            match field {
+                ("id", Value::Number(id_value)) if id.is_none() => {
+                    id = Some(id_value.as_u64().unwrap())
+                }
+                ("name", Value::String(name_value)) if name.is_none() => name = Some(name_value),
+                ("duration", Value::Number(duration_value)) if duration.is_none() => {
+                    duration = Some(Duration::from_nanos(duration_value.as_u64().unwrap()))
```

Comment:
Same string values are used here in two functions, better to move them into consts

Reply:
Btw, i would prefer to make a derive macro for that, as telemetry will grow.